File: storage/gorm/match_repo.go
// Package gorm provides GORM-based repository implementations.
package gorm

import (
	"context"
	"time"

	"gorm.io/gorm"
	"gorm.io/gorm/clause"

	"pharmabroker/domain/entity"
	"pharmabroker/domain/repository"
)

// Compile-time check that MatchRepo implements the interface
var _ repository.MatchRepository = (*MatchRepo)(nil)

// MatchRepo implements repository.MatchRepository using GORM
type MatchRepo struct {
	db *DB
}

// NewMatchRepo creates a new GORM-based match repository
func NewMatchRepo(db *DB) *MatchRepo {
	return &MatchRepo{db: db}
}

// Save creates or updates a match (upsert on offer_id + request_id)
func (r *MatchRepo) Save(ctx context.Context, match *entity.Match) error {
	model := ToMatchModel(match)

	return r.db.Conn.WithContext(ctx).
		Clauses(clause.OnConflict{
			Columns: []clause.Column{{Name: "offer_id"}, {Name: "request_id"}},
			DoUpdates: clause.AssignmentColumns([]string{
				"score", "reasoning",
			}),
		}).
		Create(model).Error
}

// GetByID retrieves a match by its ID
func (r *MatchRepo) GetByID(ctx context.Context, id string) (*entity.Match, error) {
	var model Match
	err := r.db.Conn.WithContext(ctx).
		Where("id = ?", id).
		First(&model).Error
	if err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, nil
		}
		return nil, err
	}
	return ToMatchEntity(&model), nil
}

// GetPending retrieves pending matches with full offer and request details
func (r *MatchRepo) GetPending(ctx context.Context, limit, offset int) ([]*entity.MatchWithDetails, error) {
	var matches []Match
	err := r.db.Conn.WithContext(ctx).
		Preload("Offer").
		Preload("Request").
		Where("status = ?", "PENDING").
		Order("score DESC, created_at DESC").
		Limit(limit).
		Offset(offset).
		Find(&matches).Error
	if err != nil {
		return nil, err
	}
	return ToMatchesWithDetailsEntity(matches), nil
}

// GetByOfferID retrieves all matches for a given offer
func (r *MatchRepo) GetByOfferID(ctx context.Context, offerID string) ([]*entity.Match, error) {
	var matches []Match
	err := r.db.Conn.WithContext(ctx).
		Where("offer_id = ?", offerID).
		Find(&matches).Error
	if err != nil {
		return nil, err
	}
	return ToMatchesEntity(matches), nil
}

// GetByRequestID retrieves all matches for a given request
func (r *MatchRepo) GetByRequestID(ctx context.Context, requestID string) ([]*entity.Match, error) {
	var matches []Match
	err := r.db.Conn.WithContext(ctx).
		Where("request_id = ?", requestID).
		Find(&matches).Error
	if err != nil {
		return nil, err
	}
	return ToMatchesEntity(matches), nil
}

// UpdateStatus updates the status of a match
func (r *MatchRepo) UpdateStatus(ctx context.Context, id string, status entity.MatchStatus, matchedBy, notes string) error {
	updates := map[string]interface{}{
		"status":     string(status),
		"matched_by": matchedBy,
		"notes":      notes,
	}

	if status == entity.MatchStatusConfirmed {
		updates["confirmed_at"] = time.Now()
	}

	return r.db.Conn.WithContext(ctx).
		Model(&Match{}).
		Where("id = ?", id).
		Updates(updates).Error
}

// CountPending returns the count of pending matches
func (r *MatchRepo) CountPending(ctx context.Context) (int64, error) {
	var count int64
	err := r.db.Conn.WithContext(ctx).
		Model(&Match{}).
		Where("status = ?", "PENDING").
		Count(&count).Error
	return count, err
}

// CountConfirmedToday returns the count of matches confirmed today
func (r *MatchRepo) CountConfirmedToday(ctx context.Context) (int64, error) {
	var count int64
	err := r.db.Conn.WithContext(ctx).
		Model(&Match{}).
		Where("status = ? AND date(confirmed_at) = date('now')", "CONFIRMED").
		Count(&count).Error
	return count, err
}

// GetStaleMatches finds matches older than maxAge with specified statuses (for escalation)
func (r *MatchRepo) GetStaleMatches(ctx context.Context, statuses []entity.MatchStatus, maxAge time.Duration, limit int) ([]*entity.Match, error) {
	if len(statuses) == 0 {
		return nil, nil
	}

	// Convert statuses to strings
	statusStrings := make([]string, len(statuses))
	for i, s := range statuses {
		statusStrings[i] = string(s)
	}

	cutoff := time.Now().Add(-maxAge)

	var matches []Match
	err := r.db.Conn.WithContext(ctx).
		Where("status IN ?", statusStrings).
		Where("created_at < ?", cutoff).
		Where("confirmed_at IS NULL"). // Only unresolved matches
		Order("created_at ASC").       // Oldest first
		Limit(limit).
		Find(&matches).Error
	if err != nil {
		return nil, err
	}

	return ToMatchesEntity(matches), nil
}

File: core/src/repository/postgres/match_repo.rs
//! PostgreSQL Match Repository (runtime queries)

use async_trait::async_trait;
use sqlx::{PgPool, Row};

use crate::domain::{Match, MatchStatus};
use crate::repository::MatchRepository;
use crate::{Error, Result};

pub struct PostgresMatchRepo {
    pool: PgPool,
}

impl PostgresMatchRepo {
    pub fn new(pool: PgPool) -> Self {
        Self { pool }
    }
}

#[async_trait]
impl MatchRepository for PostgresMatchRepo {
    async fn get_by_id(&self, id: &str) -> Result<Option<Match>> {
        let m = sqlx::query_as::<_, Match>(
            r#"SELECT id, offer_id, request_id, score, reasoning, matched_by,
                      status, created_at, confirmed_at, notes
               FROM matches WHERE id = $1"#,
        )
        .bind(id)
        .fetch_optional(&self.pool)
        .await?;
        Ok(m)
    }

    async fn get_pending(&self, limit: i64, offset: i64) -> Result<Vec<Match>> {
        let matches = sqlx::query_as::<_, Match>(
            r#"SELECT id, offer_id, request_id, score, reasoning, matched_by,
                      status, created_at, confirmed_at, notes
               FROM matches WHERE status = 'PENDING'
               ORDER BY score DESC, created_at DESC LIMIT $1 OFFSET $2"#,
        )
        .bind(limit)
        .bind(offset)
        .fetch_all(&self.pool)
        .await?;
        Ok(matches)
    }

    async fn count_pending(&self) -> Result<i64> {
        let row = sqlx::query("SELECT COUNT(*) as count FROM matches WHERE status = 'PENDING'")
            .fetch_one(&self.pool)
            .await?;
        Ok(row.get::<i64, _>("count"))
    }

    async fn exists(&self, offer_id: &str, request_id: &str) -> Result<bool> {
        let row = sqlx::query(
            "SELECT EXISTS(SELECT 1 FROM matches WHERE offer_id = $1 AND request_id = $2) as ex",
        )
        .bind(offer_id)
        .bind(request_id)
        .fetch_one(&self.pool)
        .await?;
        Ok(row.get::<bool, _>("ex"))
    }

    async fn save(&self, m: &Match) -> Result<()> {
        sqlx::query(
            r#"INSERT INTO matches (id, offer_id, request_id, score, reasoning, matched_by,
                status, created_at, confirmed_at, notes)
               VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10)
               ON CONFLICT (id) DO UPDATE SET score = EXCLUDED.score, reasoning = EXCLUDED.reasoning,
                 status = EXCLUDED.status, confirmed_at = EXCLUDED.confirmed_at, notes = EXCLUDED.notes"#
        )
        .bind(&m.id).bind(&m.offer_id).bind(&m.request_id).bind(m.score)
        .bind(&m.reasoning).bind(&m.matched_by).bind(m.status.to_string())
        .bind(m.created_at).bind(m.confirmed_at).bind(&m.notes)
        .execute(&self.pool).await?;
        Ok(())
    }

    async fn update_status(
        &self,
        id: &str,
        status: MatchStatus,
        matched_by: &str,
        notes: &str,
    ) -> Result<()> {
        let confirmed_at = if status == MatchStatus::Confirmed {
            Some(chrono::Utc::now())
        } else {
            None
        };
        let result = sqlx::query("UPDATE matches SET status = $1, matched_by = $2, notes = $3, confirmed_at = $4 WHERE id = $5")
            .bind(status.to_string()).bind(matched_by).bind(notes).bind(confirmed_at).bind(id)
            .execute(&self.pool).await?;
        if result.rows_affected() == 0 {
            return Err(Error::not_found(format!("Match {}", id)));
        }
        Ok(())
    }

    async fn delete_before(&self, cutoff: &chrono::DateTime<chrono::Utc>) -> Result<usize> {
        let result = sqlx::query("DELETE FROM matches WHERE created_at < $1")
            .bind(cutoff)
            .execute(&self.pool)
            .await?;
        Ok(result.rows_affected() as usize)
    }
}
